File: ios/Sources/VadCutIOS/SileroVadEngine.swift
import CryptoKit
import Foundation
import onnxruntime_objc

final class SileroVadEngine: ActivityDetector {
    static let modelSHA256 = "1a153a22f4509e292a94e67d6f9b85e8deb25b4988682b7e174c65279d8788e3"

    private let environment: ORTEnv
    private let session: ORTSession
    private var recurrentState = Array(repeating: Float(0), count: 256)
    private var contextWindow = Array(repeating: Float(0), count: 64)

    init(verifyIntegrity: Bool) throws {
        let modelURL = try VadCutResources.modelURL()
        if verifyIntegrity {
            do {
                let data = try Data(contentsOf: modelURL, options: .mappedIfSafe)
                let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
                guard digest.caseInsensitiveCompare(Self.modelSHA256) == .orderedSame else {
                    throw TrimError(
                        code: .modelIntegrityFailed,
                        message: "Bundled Silero VAD model failed its SHA-256 integrity check"
                    )
                }
            } catch let error as TrimError {
                throw error
            } catch {
                throw TrimError(
                    code: .modelLoadFailed,
                    message: "Unable to read the bundled Silero VAD model",
                    underlying: error
                )
            }
        }

        do {
            environment = try ORTEnv(loggingLevel: .warning)
            let options = try ORTSessionOptions()
            try options.setIntraOpNumThreads(1)
            session = try ORTSession(
                env: environment,
                modelPath: modelURL.path,
                sessionOptions: options
            )
        } catch {
            throw TrimError(
                code: .modelLoadFailed,
                message: "Unable to initialize ONNX Runtime for Silero VAD",
                underlying: error
            )
        }
    }

    func score(samples: [Float], validSampleCount: Int) throws -> Float {
        guard samples.count >= VadFrameCollector.frameSampleCount else {
            throw TrimError(code: .analysisFailed, message: "Silero received an incomplete frame buffer")
        }

        let modelInput = contextWindow + Array(samples.prefix(VadFrameCollector.frameSampleCount))
        let sampleRate: [Int64] = [Int64(VadFrameCollector.sampleRate)]

        do {
            let inputValue = try ORTValue(
                tensorData: NSMutableData(data: Data(copyingBufferOf: modelInput)),
                elementType: .float,
                shape: [1, 576]
            )
            let stateValue = try ORTValue(
                tensorData: NSMutableData(data: Data(copyingBufferOf: recurrentState)),
                elementType: .float,
                shape: [2, 1, 128]
            )
            let sampleRateValue = try ORTValue(
                tensorData: NSMutableData(data: Data(copyingBufferOf: sampleRate)),
                elementType: .int64,
                shape: [1]
            )

            let outputs = try session.run(
                withInputs: [
                    "input": inputValue,
                    "state": stateValue,
                    "sr": sampleRateValue,
                ],
                outputNames: ["output", "stateN"],
                runOptions: nil
            )
            guard let probabilityValue = outputs["output"],
                  let stateOutput = outputs["stateN"],
                  let probability: Float = (try probabilityValue.tensorData() as Data).copiedArray().first else {
                throw TrimError(code: .analysisFailed, message: "Silero returned incomplete outputs")
            }
            let nextState: [Float] = (try stateOutput.tensorData() as Data).copiedArray()
            guard nextState.count == recurrentState.count else {
                throw TrimError(code: .analysisFailed, message: "Silero returned an invalid recurrent state")
            }
            recurrentState = nextState
            contextWindow = Array(
                samples[(VadFrameCollector.frameSampleCount - contextWindow.count)..<VadFrameCollector.frameSampleCount]
            )
            return min(1, max(0, probability))
        } catch let error as TrimError {
            throw error
        } catch {
            throw TrimError(
                code: .analysisFailed,
                message: "Silero VAD inference failed",
                underlying: error
            )
        }
    }
}

private final class VadCutBundleToken {}

enum VadCutResources {
    static func modelURL() throws -> URL {
        let directBundles = [Bundle(for: VadCutBundleToken.self), Bundle.main]
        for bundle in directBundles {
            if let url = bundle.url(forResource: "silero_vad", withExtension: "onnx") {
                return url
            }
            if let resourcesURL = bundle.url(forResource: "VadCutIOSResources", withExtension: "bundle"),
               let resourcesBundle = Bundle(url: resourcesURL),
               let url = resourcesBundle.url(forResource: "silero_vad", withExtension: "onnx") {
                return url
            }
        }
        throw TrimError(code: .modelLoadFailed, message: "Bundled Silero VAD model was not found")
    }
}

private extension Data {
    init<T>(copyingBufferOf array: [T]) {
        self = array.withUnsafeBufferPointer { Data(buffer: $0) }
    }

    func copiedArray<T>() -> [T] {
        guard count % MemoryLayout<T>.stride == 0 else { return [] }
        return withUnsafeBytes { rawBuffer in
            Array(rawBuffer.bindMemory(to: T.self))
        }
    }
}

File: android/vadcut/src/main/java/com/vadcut/android/internal/SileroVadEngine.kt
package com.vadcut.android.internal

import android.content.Context
import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import com.vadcut.android.TrimErrorCode
import com.vadcut.android.TrimException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer
import java.nio.LongBuffer
import java.security.MessageDigest

internal class SileroVadEngine(
    context: Context,
    verifyIntegrity: Boolean,
) : ActivityDetector {
    private val environment = OrtEnvironment.getEnvironment()
    private val session: OrtSession
    private val inputBuffer: FloatBuffer = directFloatBuffer(INPUT_SAMPLE_COUNT)
    private val stateBuffer: FloatBuffer = directFloatBuffer(STATE_SIZE)
    private val sampleRateBuffer: LongBuffer = ByteBuffer.allocateDirect(Long.SIZE_BYTES)
        .order(ByteOrder.nativeOrder())
        .asLongBuffer()
    private val inputTensor: OnnxTensor
    private val stateTensor: OnnxTensor
    private val sampleRateTensor: OnnxTensor
    private val recurrentState = FloatArray(STATE_SIZE)
    private val contextWindow = FloatArray(CONTEXT_SAMPLE_COUNT)

    init {
        val model = try {
            context.assets.open(MODEL_ASSET_PATH).use { it.readBytes() }
        } catch (error: Throwable) {
            throw TrimException(TrimErrorCode.MODEL_LOAD_FAILED, "Unable to load the bundled Silero VAD model", error)
        }
        if (verifyIntegrity) {
            val actualHash = MessageDigest.getInstance("SHA-256")
                .digest(model)
                .joinToString("") { "%02x".format(it) }
            if (!actualHash.equals(MODEL_SHA256, ignoreCase = true)) {
                throw TrimException(
                    TrimErrorCode.MODEL_INTEGRITY_FAILED,
                    "Bundled Silero VAD model failed its SHA-256 integrity check",
                )
            }
        }

        session = try {
            OrtSession.SessionOptions().use { options ->
                options.setIntraOpNumThreads(1)
                options.setInterOpNumThreads(1)
                options.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)
                environment.createSession(model, options)
            }
        } catch (error: Throwable) {
            throw TrimException(TrimErrorCode.MODEL_LOAD_FAILED, "Unable to initialize ONNX Runtime", error)
        }

        inputTensor = OnnxTensor.createTensor(environment, inputBuffer, longArrayOf(1, INPUT_SAMPLE_COUNT.toLong()))
        stateTensor = OnnxTensor.createTensor(environment, stateBuffer, longArrayOf(2, 1, 128))
        sampleRateBuffer.put(0, SAMPLE_RATE.toLong())
        sampleRateTensor = OnnxTensor.createTensor(environment, sampleRateBuffer, longArrayOf(1))
    }

    override fun score(samples: FloatArray, validSampleCount: Int): Float {
        inputBuffer.clear()
        inputBuffer.put(contextWindow)
        inputBuffer.put(samples, 0, FRAME_SAMPLE_COUNT)
        inputBuffer.rewind()

        stateBuffer.clear()
        stateBuffer.put(recurrentState)
        stateBuffer.rewind()

        val probability = session.run(
            mapOf(
                "input" to inputTensor,
                "state" to stateTensor,
                "sr" to sampleRateTensor,
            ),
        ).use { result ->
            val output = result.get("output").orElseThrow() as OnnxTensor
            val nextState = result.get("stateN").orElseThrow() as OnnxTensor
            val nextStateBuffer = nextState.floatBuffer
            nextStateBuffer.get(recurrentState, 0, STATE_SIZE)
            output.floatBuffer.get(0)
        }

        System.arraycopy(
            samples,
            FRAME_SAMPLE_COUNT - CONTEXT_SAMPLE_COUNT,
            contextWindow,
            0,
            CONTEXT_SAMPLE_COUNT,
        )
        return probability.coerceIn(0f, 1f)
    }

    override fun close() {
        sampleRateTensor.close()
        stateTensor.close()
        inputTensor.close()
        session.close()
    }

    companion object {
        const val SAMPLE_RATE = 16_000
        const val FRAME_SAMPLE_COUNT = 512
        private const val CONTEXT_SAMPLE_COUNT = 64
        private const val INPUT_SAMPLE_COUNT = FRAME_SAMPLE_COUNT + CONTEXT_SAMPLE_COUNT
        private const val STATE_SIZE = 256
        private const val MODEL_ASSET_PATH = "vadcut/silero_vad.onnx"
        private const val MODEL_SHA256 = "1a153a22f4509e292a94e67d6f9b85e8deb25b4988682b7e174c65279d8788e3"

        private fun directFloatBuffer(capacity: Int): FloatBuffer =
            ByteBuffer.allocateDirect(capacity * Float.SIZE_BYTES)
                .order(ByteOrder.nativeOrder())
                .asFloatBuffer()
    }
}
